### ISSUE TITLE
netty is bumping this issue because they don't think it's important, but spamming every modded minecraft player with nonsense matters

### ISSUE THREAD

**Issue by MrMasochism** [MinecraftForge/MinecraftForge#1848]
When creating a minecraft server using a forge installer (note this doesn't happen in the dev environment) netty is installed at a later version than used in the dev environment.
When connected to via any client it will spam warnings out into the cmd window (if run in linux) but not into the server console.
This is the warning:
May 04, 2015 11:27:34 AM io.netty.channel.ChannelOutboundBuffer safeSuccess
WARNING: Failed to mark a promise as success because it is done already: Default
ChannelPromise@5404d5c5(success)

By upgrading netty in the dev environment these warnings are spamming out of packets sent by fml on entity spawn and entity update position packets.  I'm not sure if it happens on other packets but it's hard to see past those.

From netty's git:
https://github.com/netty/netty/issues/2302

**Comment by cpw** [comment]
In 1.7.10, we use netty 4.0.10. In 1.8, both FML and vanilla use 4.0.15. I see no references to 4.0.17 anywhere. I'm going to suggest that something else, not Forge or FML, has caused you to download 4.0.17 into your dev env.

**Comment by MrMasochism** [comment]
Have you tried it?  Closing this without investigation seems a little silly.  I have the reports of 6 other people who've attempted to set up a forge server and have had exactly the same issue

**Comment by clienthax** [comment]
Please note this only occurs when running in Dedicated server mode.

**Comment by chibill** [comment]
Maybe mojang (if 1.8) used a newer version of Netty..  Then on the client?

**Comment by LexManos** [comment]
Yes, I have. This wasn't closed without investigation.
The dedicated server bundles all dependencies including netty. Thus freezing them into the specific version unless you go along and override it. If it WAS a real issue then we would be getting hundreds of angry server provides hounding us.
However, I suspect that your users are stupid and are trying to run Forge on a 1.8.2+ server, which is where netty was changed to 4.0.23.

**Comment by MrMasochism** [comment]
I just did a fresh install of a minecraft server using the forge installer to set it up.  Got exactly the same issue.  Definitely not a 1.8.2 server, never even had a 1.8.2 install on my pc

**Comment by LexManos** [comment]
Logs, hashes, etc etc.. report your issue correctly and we might entertain debugging it.
However, it's been proven to not be an issue on anyone else's machines, and not a issue on Forge's end.
There should be no reason what so ever that netty is updated on 1.8. 1.8 ships a compatible version of netty as just tested. Track it down yourself and see if there are conflicts in your setup.

**Comment by MrMasochism** [comment]
this isn't in the dev side.  This is a fresh install of a minecraft server.  I reported it extremely clearly above.  Everyone else I know who's tried the same steps got the same issue.  It may have to have a mod that spawns entities in it running to cause the issue.  Definitely happens with pixelmon and separately with custom npcs

**Comment by clienthax** [comment]
So, lets download this https://s3.amazonaws.com/Minecraft.Download/versions/1.8/minecraft_server.1.8.jar
And then lets open it in 7z or similar,
and open up META-INF/io.netty.versions.properties
O whats that, the version number isn't 4.0.10 like cpw said it would be.
o shit me, look netty-transport-rxtx.version=4.0.20.Final

**Comment by LexManos** [comment]
Find a way to reproduce that isn't "Install these complicated mods that could be causing the issue itself"
You may also want to you know POST THE DAMN LOGS like i've asked every freakin post!

**Comment by clienthax** [comment]
Seems this is actually a mojang issue.
9:50 PM <Grum> io\netty\netty-all\4.0.10.Final\netty-all-4.0.10.Final.jar in 1.7.10
9:51 PM <clienthax> yeah, got it for 1.8 ?
9:51 PM <Grum> 1.8.0 io\netty\netty-all\4.0.15.Final\netty-all-4.0.15.Final.jar
9:51 PM <clienthax> then why is 4.0.20 in the jar at https://s3.amazonaws.com/Minecraft.Download/versions/1.8/minecraft_server.1.8.jar
9:51 PM <clienthax> thats what this whole crap is about
9:52 PM <Grum> io\netty\netty-all\4.0.23.Final\netty-all-4.0.23.Final.jar for 1.8.4
9:52 PM <clienthax> http://prntscr.com/71kg94
9:52 PM <clienthax> maybe you lot released 2 versions of 1.8 ?
9:53 PM <Grum> nope, it seems we're shipping different client and srever version, wtf :P
9:53 PM <Grum>     compile 'io.netty:netty-all:4.0.20.Final' it says for release-1.8
9:53 PM <Grum> but the clients use older for some reason

**Comment by LexManos** [comment]
It may be a mojang issue but again, things need to be reported properly and addressed/reproduced. 'This thing happens cuz I say so' doesn't help development.

**Comment by clienthax** [comment]
Fresh forge latest install, Server log -> http://pastebin.ubuntu.com/10986738/
the netty messages do not get logged to the forge log.
console log -> http://pastebin.ubuntu.com/10986740/

**Comment by clienthax** [comment]
As previously stated it seems to appear when a mod spawns a entity, (tested on pixelmon and custom npcs)

heres a test server https://mega.co.nz/#!jdcgRL4Y!GkL1fSz4sOXvmHW6Kr0TnA-JRSoVI7_xE6TEMfiHcr8

thanks for looking into this

**Comment by cpw** [comment]
OK. So this issue is in fact that mojang fucked up the dediserver jar for 1.8, or rather that they expected the client to use a newer version of netty than it does.

I'll look into how to fix the netty spamminess, but this isn't exactly something we could anticipate, and your report is NOT clear at all..

**Comment by clienthax** [comment]
So, reopening the issue?

**Comment by cpw** [comment]
Based on the netty issue you linked, there is nothing we can do, seemingly, without significantly changing some important behaviours. I'm not comfortable doing that for a mojang derp, I would rather mojang fix their netty.

**Comment by cpw** [comment]
OK. After diving through the code, it's even worse: https://github.com/netty/netty/commit/7041a9238e5087efa5fb0fa3d8a5d0721eb5f610 is the commit that caused the problem. I have no idea how FML is now triggering that, though I suspect it will happen in a very widespread manner. Clearly, the netty devs have no idea either, https://github.com/netty/netty/issues/2302 is full of sliding "we'll fix" reports.

**Comment by Bogdan-G** [comment]
Hmm, custom version Netty and update Netty 4.1 and newer? (Netty 4.0.0.25 not solved?)

**Comment by clienthax** [comment]
https://github.com/netty/netty/commit/6c094237e5c46b9122269d9a3394a4e9707e5570

**Comment by clienthax** [comment]
Could the issue be that the fml message to message packets arn't being flushed?
https://github.com/netty/netty/blob/master/codec/src/main/java/io/netty/handler/codec/MessageToMessageEncoder.java#L81

**Comment by jk-5** [comment]
You don't need to flush messages generated in a codec. It's done by the code sending that packet after the encoding is done. That part has worked for a long time so that is not the issue

**Comment by cpw** [comment]
No. The messages flush fine. The FML channel is built on top of the embedded channel. We simply write to it.. This looks to be a netty bug...

**Comment by clienthax** [comment]
@cpw How comes minecrafts implementation doesn't spam?

**Comment by cpw** [comment]
They don't use embedded channels..

**Comment by clienthax** [comment]
https://github.com/MinecraftForge/FML/commit/d89165021f33fbffb4563d86b30bd261506c6ea6

Seems if you remove those, netty will be able to mark the promise a success when it runs remove,
would that have any adverse affects?

**Comment by cpw** [comment]
It could very well. I added that so netty would present errors to the system.. I'm reluctant to remove something without very good evidence here, especially since netty have NOT closed the parent issue, indicating to me that they see the bug as well..

**Comment by LexManos** [comment]
May be worth forcing a downgrade of Netty using the installer. To make it in sync with the client, and to avoid this bug. But beyond that, it would be nice to have a small simple mod that reproduces this issue.

**Comment by clienthax** [comment]
@LexManos Test mod source https://drive.google.com/file/d/0B3htWWNJdaBpdW1oWk9oM0VKUTQ/view?usp=sharing
Test mod built https://drive.google.com/file/d/0B3htWWNJdaBpOWsweF9uQkw1NjA/view?usp=sharing

Will trigure the issue once per mod entity spawn with the "/spamtest" command

**Comment by clienthax** [comment]
Also i notice that packets that reach remove that are set to UNCANCELLABLE seem to not spam.

**Comment by cpw** [comment]
Interesting observation..

**Comment by cpw** [comment]
Update: we have to wait for netty to fix it.. We're effectively blocked by https://github.com/netty/netty/issues/2302

**Comment by JordanMajd** [comment]
I am also experiencing this issue with a clean server install on a fresh copy of Win 10 Preview x64 with forge-1.8-11.14.1.1353-universal.jar, minecraft_server.1.8.jar
```
WARNING: Failed to mark a promise as success because it is done already: DefaultChannelPromise@417711f7(success)
May 20, 2015 6:07:35 PM io.netty.channel.ChannelOutboundBuffer safeSuccess
```

**Comment by LexManos** [comment]
What is the viability of forcing a netty downgrade on the server? Our installer can install the jar just fine, we just need to make sure java uses it. Perhaps by writing a re-direction transformer that forces it to load from the jar. Or seeing if we can just make it appear before minecraft_server.jar in the classpath.

**Comment by jk-5** [comment]
Putting a `Boot-Class-Path: netty.jar` entry in the manifest will force the JVM to load netty from that jar before from the mc jar. It is a really hacky way of doing it. Not sure how it would behave with the custom ClassLoader, but It might work.

**Comment by cpw** [comment]
@jk-5 hacky, but effective. We control the forge jar manifest, so this is eminently doable.

**Comment by LexManos** [comment]
It's worth a test. and the CCL is slave to master so it shouldn't be an issue. If anyone is transforming netty they are probably doing it wrong. Worth taking a look.

**Comment by LexManos** [comment]
Tested with @clienthax's mod, could not reproduce

**Comment by LexManos** [comment]
Got it reproducible and this does shut it up. It's a nasty fix, but it works. Also has benefit of syncing libraries between client and server.

**Comment by mrgoodrich** [comment]
Could somebody possibly explain how I could simply do what you stated?  I'm not as experienced in this area, but would love to shut up the netty thing..

**Comment by clienthax** [comment]
@mrgoodrich should be fixed in the latest forge builds.

**Comment by mrgoodrich** [comment]
I'm using forge 1.8.0 due to mod restrictions and sponge, so I should just ignore it then?

**Comment by LexManos** [comment]
Just update.

**Comment by mrgoodrich** [comment]
Pixelmon and Sponge will still work with versions of forge past 1.8.0?

**Comment by mrgoodrich** [comment]
Didn't know it would work, thank you!

**Comment by mrgoodrich** [comment]
Yeahhh so I updated to the latest version of forge, and it's still spamming the following error :/

[io.netty.channel.ChannelOutboundBuffer]: Failed to mark a promise as success because it is done already: DefaultChannelPromise@6e71cdbb(success)

**Comment by LexManos** [comment]
You need to use the installer and not just replace the universal jar. Or manually place netty into the libraries folder.
Or worse case scenario, just add netty to your classpath before everything else.

**Comment by mrgoodrich** [comment]
That makes sense and worked, I was confused on the issue before.  Thanks @LexManos !

**Comment by ZombifiedTech** [comment]
What exactly did you download & do to fix this issue? Link to what installer. I'll do a fresh one and just move my world over.

**Comment by clienthax** [comment]
@LexManos Same issue 1.8.8 branch

**Comment by cpw** [comment]
@clienthax this will have to be lived with. 1.8.8 requires the bad netty version. And they show no signs of fixing the original bug.

**Comment by clienthax** [comment]
@cpw changing the version back to the one you changed it to works fine, why not just apply the same patch?

**Comment by cpw** [comment]
@clienthax, that was valid in 1.8, but 1.8.8 both client and server use 23. We're not in the business of rolling back libs for logspam.

**Comment by cpw** [comment]
@clienthax https://github.com/netty/netty/pull/2229 was accepted in 4.0.17. I know searge worked on making that work. I will check if he is actively using it in 1.8.8. Note: we can't forever roll back, since both client and server are on 23 I think we draw a line. I will dig in and see if I can fix via fml code somehow, but I'm skeptical..

**Comment by cpw** [comment]
OK. I have just verified that Epoll is used in 1.8.8. Downgrading is *not* possible. It will cause the server to fail.

**Comment by Bogdan-G** [comment]
This library Netty is a headache.

**Comment by cpw** [comment]
Reopening because the netty link shows this as closed. It is **still** an issue

**Comment by cpw** [comment]
Those experiencing this issue, try build 1626. Please report any instances of log spam from netty, either this message or any other, with that forge version. Thanks!

**Comment by Thutmose** [comment]
I can confirm that it has removed at least most of the spam from my mod, It used to happen once or so per packet sent by a mob, which is about 1 packet every few seconds, currently have about 1000 of them around, no netty spam seen yet.

**Comment by cpw** [comment]
No other errors, spam or weird net glitches?

**Comment by Thutmose** [comment]
no other errors or spam showed up in the console, I didn't notice any glitches, so nothing from 1000 wild pokemobs doing whatever they do for about 20 minutes caused anything, which should cover 90% of my packets I send (by number).  nothing seems to have gone wrong in the other stuff I have done, but I haven't done much vigorous testing on that.

**Comment by fnuecke** [comment]
In the case where it reliably happened for me (spawning custom entities) it no longer happens in 1628.
No oddities observed (only tested for a couple of minutes though, for what it's worth).

**Comment by LexManos** [comment]
Closing as this doesnt seem to be an issue anymore, the new netty is updated and synced between both client and server should be fine.
